cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll

```diff
@@ -1272,9 +1292,45 @@ private predicate flowCandFwdReadTaint(
 ) {
   exists(Node mid, AccessPathFront apf |
     flowCandFwd(mid, fromArg, apf, config) and
-    argumentFlowsThrough(mid, node, _, t, TSummaryReadTaint(f), config) and
-    apf.headUsesContent(f) and
-    nodeCand(node, unbind(config))
+    argumentFlowsThrough2(mid, node, _, t, TSummaryReadTaint(f), config) and
+    apf.headUsesContent(f)
+  )
+}
+
+pragma[noinline]
+private predicate flowCandFwdEmptyAp(Node node, Configuration config) {
```

Comment:
Does it make sense to have
```
pragma[noinline]
private predicate flowCandFwdNonEmptyAp(Node node, Content f, Configuration config) {
  flowCandFwd(node, _, any(AccessPathFront apf | apf.headUsesContent(f)), config)
}
```
as well, and then use that further down?

Reply:
It appears to be fast enough as-is, so I don't think it's needed.  Note that the empty accesspath case is different as we're projecting away the type in `flowCandFwdEmptyAp`, which is why that is factored out.